src/Middleware/Session/src/SessionMiddleware.cs

```diff
@@ -104,7 +104,15 @@ public async Task Invoke(HttpContext context)
 
             try
             {
-                await _next(context);
+                if (_logger.IsEnabled(LogLevel.Critical))
+                {
+                    using var scope = _logger.BeginScope(new SessionLogScope(feature.Session.Id));
```

Comment:
Accessing `Session.Id` will call `ISession.Load` and `LoadAsync` would never do anything because `_isLoaded` is always true.
https://github.com/dotnet/aspnetcore/blob/0cf96e6749cf83269f57e216da9bbd3637e0f000/src/Middleware/Session/src/DistributedSession.cs#L111-L121
https://github.com/dotnet/aspnetcore/blob/0cf96e6749cf83269f57e216da9bbd3637e0f000/src/Middleware/Session/src/DistributedSession.cs#L231-L235
Is this an issue? When `LoadAsync` is expected to be called?

This is also the reason that the tests are failing. Should I change the tests for `LoadAsync` so it is being tested directly from `DistributedSession` and not middleware pipeline?

Reply:
Session is loaded on first access which includes accessing the session id. That means session won't be loaded on a request if it's never accessed. We can't eagerly access session id to create this logger scope because it would cause session to be loaded on all requests.

I don't think we can do this feature.

Reply:
What if we use `sessionKey` instead of `Session.Id`?
https://github.com/dotnet/aspnetcore/blob/7b84159406041ddd38658b2ea07058cbb8a0f20a/src/Middleware/Session/src/SessionMiddleware.cs#L92-L94
https://github.com/dotnet/aspnetcore/blob/7b84159406041ddd38658b2ea07058cbb8a0f20a/src/Middleware/Session/src/DistributedSession.cs#L131-L132
They are both equally meaningless! (at least in the default implementation) and in this case there's no need to eagerly load the session and we would have session scope in all requests whether they access session or not, which is a good thing for checking logs for a session.

Reply:
sessionKey is even more meaningless than the session id, these's no guarantee of an actual session behind it, and it can be re-used across multiple sessions. The session key is also only lazily created if session id used.

Reply:
https://github.com/dotnet/aspnetcore/blob/7b84159406041ddd38658b2ea07058cbb8a0f20a/src/Middleware/Session/src/DistributedSession.cs#L172
https://github.com/dotnet/aspnetcore/blob/7b84159406041ddd38658b2ea07058cbb8a0f20a/src/Middleware/Session/src/SessionMiddleware.cs#L157

Oh I see! The value for cookie is created but the cookie itself is only added to the response if `Session.Set()` is called. 
I'll open another issue to talk about if the extra allocation of a byte[] and two strings could be prevented.

>  it can be re-used across multiple sessions

But are those sessions for the same cookie value and in the same browser? If so in the context of request correlation that's not so bad.

- How about only create a session scope if sessionKey is available in the cookie? A major downside would be that the first request does not contains the scope.

- Or Introduce a new `ISession.IsLoaded` property and pass `ISession` into the log scope and only return session id if `IsLoaded` is true.


- At this point I think I don't really care about session! A simple protected cookie value would also do the trick. Any Idea?
```csharp
public class CorrelationScopeMiddleware
{
    public async Task Invoke(HttpContext context)
    {
        var cookieValue = context.Request.Cookies[_options.Cookie.Name!];
        var correlationValue = CookieProtection.Unprotect(_dataProtector, cookieValue, _logger);

        if (string.IsNullOrWhiteSpace(correlationValue)
        {
            var guidBytes = new byte[16];
            RandomNumberGenerator.Fill(guidBytes);
            correlationValue = new Guid(guidBytes).ToString();
            cookieValue = CookieProtection.Protect(_dataProtector, correlationValue);
            context.Response.OnStarting(SetCookie, state: (this, cookieValue))
        }

        using var scope = new CorrelationScope(correlationValue)
        await _next(context);
    }
}

```

Reply:
> * Or Introduce a new `ISession.IsLoaded` property and pass `ISession` into the log scope and only return session id if `IsLoaded` is true.

We really should add ISession.IsLoaded at some point, but I don't think it will help in this case since session can't be loaded before the middleware runs anyways, only after.

> * At this point I think I don't really care about session! A simple protected cookie value would also do the trick. Any Idea?

Seems reasonable. Why not add the response cookie immediately? You shouldn't need the OnStarting callback for that.

Reply:
> We really should add ISession.IsLoaded at some point, but I don't think it will help in this case since session can't be loaded before the middleware runs anyways, only after.

but at some point it will loaded and after that all the logs would have the session id if we pass the `ISession` to the log scope and always check `IsLoaded` in `GetEnumerator` method.

> Seems reasonable. 

Do you think this middleware is something that ASP.NET could provide?

> Why not add the response cookie immediately? You shouldn't need the OnStarting callback for that.

You're right, I was just copying from `SessionMiddleware`

Reply:
> but at some point it will loaded and after that all the logs would have the session id if we pass the `ISession` to the log scope and always check `IsLoaded` in `GetEnumerator` method.

That seems too inconsistent to be useful? It would miss most of the framework logs and only kick in during application logic.

> Do you think this middleware is something that ASP.NET could provide?

I doubt it, we haven't seen a lot of demand for this kind of session scoping. I'd expect people to correlate using stronger signals like client IP or user id.

Reply:
Ok, so I close this PR for now. Should I also close the related issue?

Reply:
Yea, I'll take care of that now.